### ISSUE TITLE
whole-genome SNPs - LD matrix

### ISSUE THREAD

**Issue by clairemerot** [petrikemppainen/LDna#10]
Hello,
Do you think it is possible to use LDna on a large number of SNPs (>1 000 000) obtained through Whole-genome sequencing? If so what would you recommend to build the LD matrix?
Thanks 
Claire

**Comment by petrikemppainen** [comment]
Hi Claire,

Using LDna for very large data sets is not possible/practical/useful. For LDna to be useful a considerable proportion (of any total number of loci) need to be in LD, otherwise no clustering will be present. Thus, interesting patterns (if any) will also show up in subsets of the data (3000-5000 loci), and there is no need to use the full data set. If you are interested in inversions, then you can do LDna chromosome vise.

There is also the possibility to pre cluster loci in non-overlapping windows along chromosomes to remove redundant loci (that are in LD due to physical proximity only), and then do the LDna on the reduce data set for all chromosomes.

The paper describing this has just been accepted for publication in Molecular Ecology Resourcs, and I’m working on the new version of LDna that has this ‘LDnClustering’ function. The development version is .64, which you can find on https://github.com/petrikemppainen/LDna/tree/v.64

While the documentation is still laking there is some example code that you can use for the function  ‘LDnClustering’. Please report if you find any bugs and have suggestions of improvements, this is still very new.

This function produces (among others) a list of clusters (sets of loci) and if you set LD threshold settings high (say 0.6 and 0.8) then it is possible to, from each cluster, select one locus that has the highest pair vise median LD among the loci in the cluster, and have this represent the whole cluster in the global analysis. How useful this pre-clustering will be depends on the SNP density and the effective population size. This pre-clustering should be done in populations where there is not much population structure (i.e. other sources of LD than physical proximity).

The manuscript describing this window based pre-clustering of loci will appear in pre print soon.

/Petri

**Comment by clairemerot** [comment]
Thank you very much for your fast reply and the interesting suggestions.
I found your manuscript on MolEcol Ressources, I'll look into that and try v64. 
Thanks a lot
Claire

**Comment by petrikemppainen** [comment]
Yes, the e-publication is available here: https://www.ncbi.nlm.nih.gov/pubmed/29673105

We have so far used it for up up to 200Kbp and it takes about 10 minutes with the new implementation.   I am writing now an implementation where I use vcftools to estimate LD and only read in data for each window at a time from separate vcf files for each chromosome (as you know, R is very slow with large data sets). I recommend that you try out v.64 for a subset of data to see how much clustering you get. I need to analyse whole genome data sets myself now, so I think the vcf version will be done next week. 

I will also shortly implement functions where the user can choose to pick one SNP to represent the whole cluster or take the consensus sequence for the cluster (rather than using the PCs) for downstream analyses. In particular if you are using LDna you need genotype data not PCs.

Petri

**Comment by clairemerot** [comment]
Thanks Petri, 
I'll try with the subset and I'll be in touch.
To calculate LD from vcf on each chromosome, you may be interested in this new tool https://github.com/statgen/emeraLD. It works really fast.
Cheers
Claire

**Comment by petrikemppainen** [comment]
Great, will check this out!
/Petri

**Comment by petrikemppainen** [comment]
Hi again, 

I have now modified the code such that the locus with highest intra cluster median LD and the consensus SNP is exported from each cluster. This data set can be analysed with any GWAS method. I've also added the function LDnClusteringVCF which takes .vcf files as imput (and calls vcftools to estimate LD etc.). Documentation is not complete for this function, but please try and report any bugs!!

/Petri

**Comment by clairemerot** [comment]
Amazing, thanks a lot! 
I hope to have the time to try that next week. I'll keep you informed.
Thanks
Claire

**Comment by clairemerot** [comment]
sorry it took me so long to explore the new functions.
More generally, it did not work for my data. I think it is because my vcf is based on geneotype likelihood (export by Angsd because I have low-coverage data), and calculating ld requires genotypes GT. I am looking if there is a way of transforme vcf GL into vcf GT.
I'll keep working on that, maybe with plink or susbet of genotype in 0,1,2 format for LDclustering function.I'll keep you in touch

There were also a few little issue that I had at the beginning of the function that it is worth mentionning maybe?.
The function makes an error
 "Error in fread(paste0("./", temp[grep("ldchr", temp)])) : 
  File './' is a directory. Not yet implemented."

so I ran it line by line, 
I had to remove [to_do] at line 4
than I changed 
 system(paste0("vcftools --vcf ", vcf_folder, "/",vcf_files[[i]], " ", " --max-alleles 2 --min-alleles 2 --recode --recode-INFO-all"), ignore.stderr = FALSE)
(by adding the "/" between folder path and vcf file name) -> I think this was the reason for the first error which prevent vcftools to find the file.
(removing maf filter but this is specific to my vcf GL format)
(choosing ignore.stderr=FALSE to have feed-back from vcftools)

Following lines run well until hotspots, which is empty... because my LDmat is all NA as is the vcf of LD. vcftools has not been able to calculate R² on my data. I guess with regular vcf it would work.

Cheers
Claire

**Comment by petrikemppainen** [comment]
Yes, you need genotypes not genotype likelihoods. In any case I have found some bugs as well, and overall not entirely done with the code as I have been on holiday/conference for 2 weeks.

Will try to sort it out within a couple of days.

/Petri

**Comment by petrikemppainen** [comment]
Just updated v.64, works with my data, but still expect some bugs. Once you have .rds files for all chromosomes you can concatenate this into a single file using function: 'Concat_files'.
